plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,842 @@
+
+    def _validate_run_inputs(
+        self, stop_condition, save_routine, dt, field_weighting: str
+    ):
+        if not isinstance(stop_condition, AbstractStopCondition):
+            raise TypeError("Please specify a valid stop condition.")
+
+        if not isinstance(save_routine, AbstractSaveRoutine):
+            raise TypeError("Please specify a valid save routine")
+
+        require_uniform_time = (
+            stop_condition.require_uniform_dt or save_routine.require_uniform_dt
+        )
+
+        # Will the simulation follow a uniform time step?
+        # This must be the case for certain stopping conditions and saving routines
+        self.is_uniform_time = isinstance(dt, u.Quantity) and isinstance(
+            dt.value, float
+        )
+
+        # Raise a ValueError if dt is required by stop condition or save routine but not specified
+        if require_uniform_time and not self.is_uniform_time:
+            raise ValueError(
+                "Please specify a uniform time step to use the simulation with this configuration!"
+            )
+
+        # Load and validate inputs
+        field_weightings = ["volume averaged", "nearest neighbor"]
+        if field_weighting in field_weightings:
+            self.field_weighting = field_weighting
+        else:
+            raise ValueError(
+                f"{field_weighting} is not a valid option for ",
+                "field_weighting. Valid choices are",
+                f"{field_weightings}",
+            )
+
+    def _tracked_particle_mask(self):
+        """
+        Calculates a boolean mask corresponding to particles that have not been stopped or removed.
+        """
+        return ~np.logical_or(np.isnan(self.x[:, 0]), np.isnan(self.v[:, 0]))
+
+    @property
+    def nparticles_tracked(self):
+        """Return the number of particles that don't have NaN position or velocity."""
+        return self._tracked_particle_mask().sum()
+
+    def run(
+        self,
+        stop_condition: AbstractStopCondition,
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        field_weighting="volume averaged",
+    ):
+        r"""
+        Runs a particle-tracing simulation.
+        Time steps are adaptively calculated based on the
+        local grid resolution of the particles and the electric and magnetic
+        fields they are experiencing.
+
+        Parameters
+        ----------
+        stop_condition : `~plasmapy.simulation.particle_tracker.AbstractStopCondition`
+            A class responsible for determining when the simulation has reached
+            a suitable termination condition. The class is passed an instance
+            of `~plasmapy.simulation.particle_tracker.ParticleTracker`. See `~plasmapy.simulation.particle_tracker.AbstractStopCondition`
+            for more details.
+
+        save_routine: `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine`
```

Comment:
I would word the description here like this

An instance of `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine` which determines which timesteps of the simulation to save. 

That makes it clear that what is being passed in is an instance of that class, not a reference to the class itself.